### ISSUE TITLE
Named container not found (when using aufs)

### ISSUE THREAD

**Issue by akafred** [docker/docker#13319]
After creating a named container it is not found using that name (when using aufs):

From /var/log/upstart/docker.log:

Creation:

```
INFO[0012] POST /v1.18/containers/create?name=koha_mysql_container 
INFO[0012] +job create(koha_mysql_container)            
INFO[0012] +job log(create, 06ab488fcc8f8d12f8d9142b59e9f9c38b4b832309781459e998d693c3f57cbc, mysql:5.6.21) 
INFO[0012] -job log(create, 06ab488fcc8f8d12f8d9142b59e9f9c38b4b832309781459e998d693c3f57cbc, mysql:5.6.21) = OK (0) 
INFO[0012] -job create(koha_mysql_container) = OK (0) 
```

Lookup by id is ok: 

```
INFO[0012] GET /v1.18/containers/06ab488fcc8f8d12f8d9142b59e9f9c38b4b832309781459e998d693c3f57cbc/json 
INFO[0012] +job container_inspect(06ab488fcc8f8d12f8d9142b59e9f9c38b4b832309781459e998d693c3f57cbc) 
INFO[0012] -job container_inspect(06ab488fcc8f8d12f8d9142b59e9f9c38b4b832309781459e998d693c3f57cbc) = OK (0) 
```

Lookup by name *fails*:

```
INFO[0012] GET /v1.18/containers/koha_mysql_container/json 
INFO[0012] +job container_inspect(koha_mysql_container) 
no such id: koha_mysql_container
INFO[0012] -job container_inspect(koha_mysql_container) = ERR (1) 
ERRO[0012] Handler for GET /containers/{name:.*}/json returned error: no such id: koha_mysql_container 
ERRO[0012] HTTP Error: statusCode=404 no such id: koha_mysql_container 
```

(The commands are run automated using docker-py via SaltStack - so in quite rapid succession.)

`docker version`:

```
Client version: 1.6.2
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 7c8fca2
OS/Arch (client): linux/amd64
Server version: 1.6.2
Server API version: 1.18
Go version (server): go1.4.2
Git commit (server): 7c8fca2
OS/Arch (server): linux/amd64
```

`docker info`:

```
Containers: 37
Images: 211
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 285
 Dirperm1 Supported: false
Execution Driver: native-0.2
Kernel Version: 3.13.0-32-generic
Operating System: Ubuntu 14.04.2 LTS
CPUs: 4
Total Memory: 5.691 GiB
Name: dev-ship
ID: AFFL:NRE3:6M7N:7YXX:ANNC:GNRU:DLSO:V5NJ:5IU5:UXEF:RQQF:ADUH
Username: deichman
Registry: [https://index.docker.io/v1/]
WARNING: No swap limit support
```

`uname -a`:

`Linux dev-ship 3.13.0-32-generic #57-Ubuntu SMP Tue Jul 15 03:51:08 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux`

Environment details (AWS, VirtualBox, physical, etc.):

Vagrant virtualbox using Vagrant and SaltStack for provisioning.

```
Virtualbox 4.3.18
Vagrant 1.7.2
Vagrant base box: ubuntu/trusty64
SaltStack v2014.7.0
Docker-py 1.1.0
```

How reproducible:

In our setup it is quite reliably reproduced when using aufs. (With devicemapper we occationally have issues with mounting, which is why we are trying out aufs.) We haven't tried to create a minimal setup to reproduce, but if that is needed we probably can.

Actual Results:

From /var/log/upstart/docker.log - using aufs (I have omitted some GET /v1.17/version requests):

```
INFO[0012] GET /v1.18/containers/koha_mysql_container/json 
INFO[0012] +job container_inspect(koha_mysql_container) 
no such id: koha_mysql_container
INFO[0012] -job container_inspect(koha_mysql_container) = ERR (1) 
ERRO[0012] Handler for GET /containers/{name:.*}/json returned error: no such id: koha_mysql_container 
ERRO[0012] HTTP Error: statusCode=404 no such id: koha_mysql_container 
INFO[0012] POST /v1.18/containers/create?name=koha_mysql_container 
INFO[0012] +job create(koha_mysql_container)            
INFO[0012] +job log(create, 06ab488fcc8f8d12f8d9142b59e9f9c38b4b832309781459e998d693c3f57cbc, mysql:5.6.21) 
INFO[0012] -job log(create, 06ab488fcc8f8d12f8d9142b59e9f9c38b4b832309781459e998d693c3f57cbc, mysql:5.6.21) = OK (0) 
INFO[0012] -job create(koha_mysql_container) = OK (0)   
INFO[0012] GET /v1.18/containers/06ab488fcc8f8d12f8d9142b59e9f9c38b4b832309781459e998d693c3f57cbc/json 
INFO[0012] +job container_inspect(06ab488fcc8f8d12f8d9142b59e9f9c38b4b832309781459e998d693c3f57cbc) 
INFO[0012] -job container_inspect(06ab488fcc8f8d12f8d9142b59e9f9c38b4b832309781459e998d693c3f57cbc) = OK (0) 
INFO[0012] GET /v1.18/containers/koha_mysql_container/json 
INFO[0012] +job container_inspect(koha_mysql_container) 
no such id: koha_mysql_container
INFO[0012] -job container_inspect(koha_mysql_container) = ERR (1) 
ERRO[0012] Handler for GET /containers/{name:.*}/json returned error: no such id: koha_mysql_container 
ERRO[0012] HTTP Error: statusCode=404 no such id: koha_mysql_container 
```

Expected Results:

From /var/log/upstart/docker.log when using devicemapper:

```
INFO[0015] GET /v1.18/containers/koha_mysql_container/json 
INFO[0015] +job container_inspect(koha_mysql_container) 
no such id: koha_mysql_container
INFO[0015] -job container_inspect(koha_mysql_container) = ERR (1) 
ERRO[0015] Handler for GET /containers/{name:.*}/json returned error: no such id: koha_mysql_container 
ERRO[0015] HTTP Error: statusCode=404 no such id: koha_mysql_container 
INFO[0015] POST /v1.18/containers/create?name=koha_mysql_container 
INFO[0015] +job create(koha_mysql_container)            
INFO[0015] +job log(create, 95a098348251b5a8f5295545410d8e8b8ae104f25b00b8a98649eb0315be9760, mysql:5.6.21) 
INFO[0015] -job log(create, 95a098348251b5a8f5295545410d8e8b8ae104f25b00b8a98649eb0315be9760, mysql:5.6.21) = OK (0) 
INFO[0015] -job create(koha_mysql_container) = OK (0)   
INFO[0015] GET /v1.18/containers/95a098348251b5a8f5295545410d8e8b8ae104f25b00b8a98649eb0315be9760/json 
INFO[0015] +job container_inspect(95a098348251b5a8f5295545410d8e8b8ae104f25b00b8a98649eb0315be9760) 
INFO[0015] -job container_inspect(95a098348251b5a8f5295545410d8e8b8ae104f25b00b8a98649eb0315be9760) = OK (0) 
INFO[0015] GET /v1.18/containers/koha_mysql_container/json 
INFO[0015] +job container_inspect(koha_mysql_container) 
INFO[0015] -job container_inspect(koha_mysql_container) = OK (0) 
```

**Comment by thaJeztah** [comment]
Hm, funny think is that according to the docs, that endpoint of the API should only accept an ID, not a name;
https://docs.docker.com/reference/api/docker_remote_api_v1.18/#inspect-a-container, so it looks like the docs need an update here as well

However, the logs indeed show this;
```
INFO[2758214] GET /v1.18/containers/foobar/json
INFO[2758214] +job container_inspect(foobar)
INFO[2758214] -job container_inspect(foobar) = OK (0)
```

And for a non-existing container;
```
INFO[2758462] -job container_inspect(asdasdadsda) = ERR (1)
ERRO[2758462] Handler for GET /containers/{name:.*}/json returned error: no such id: asdasdadsda
ERRO[2758462] HTTP Error: statusCode=404 no such id: asdasdadsda
INFO[2758462] GET /v1.18/images/asdasdadsda/json
INFO[2758462] +job image_inspect(asdasdadsda)
No such image: asdasdadsda
INFO[2758462] -job image_inspect(asdasdadsda) = ERR (1)
ERRO[2758462] Handler for GET /images/{name:.*}/json returned error: No such image: asdasdadsda
ERRO[2758462] HTTP Error: statusCode=404 No such image: asdasdadsda
```

I'm unable to reproduce this on my machine (comparable configuration, but a newer kernel; `3.13.0-45-generic`)

Perhaps you can check your configuration using this script;
https://github.com/docker/docker/blob/master/contrib/check-config.sh

**Comment by akafred** [comment]
check-config didn't reveal anything sinister... (just stuff under optional and regarding unused storagedrivers): 
```
warning: /proc/config.gz does not exist, searching other paths for kernel config...
info: reading kernel config from /boot/config-3.13.0-32-generic ...

Generally Necessary:
- cgroup hierarchy: properly mounted [/sys/fs/cgroup]
- apparmor: enabled and tools installed
- CONFIG_NAMESPACES: enabled
- CONFIG_NET_NS: enabled
- CONFIG_PID_NS: enabled
- CONFIG_IPC_NS: enabled
- CONFIG_UTS_NS: enabled
- CONFIG_DEVPTS_MULTIPLE_INSTANCES: enabled
- CONFIG_CGROUPS: enabled
- CONFIG_CGROUP_CPUACCT: enabled
- CONFIG_CGROUP_DEVICE: enabled
- CONFIG_CGROUP_FREEZER: enabled
- CONFIG_CGROUP_SCHED: enabled
- CONFIG_CPUSETS: enabled
- CONFIG_MACVLAN: enabled
- CONFIG_VETH: enabled
- CONFIG_BRIDGE: enabled
- CONFIG_NF_NAT_IPV4: enabled
- CONFIG_IP_NF_FILTER: enabled
- CONFIG_IP_NF_TARGET_MASQUERADE: enabled
- CONFIG_NETFILTER_XT_MATCH_ADDRTYPE: enabled
- CONFIG_NETFILTER_XT_MATCH_CONNTRACK: enabled
- CONFIG_NF_NAT: enabled
- CONFIG_NF_NAT_NEEDED: enabled
- CONFIG_POSIX_MQUEUE: enabled

Optional Features:
- CONFIG_MEMCG_SWAP: enabled
- CONFIG_MEMCG_SWAP_ENABLED: missing
    (note that cgroup swap accounting is not enabled in your kernel config, you can enable it by setting boot option "swapaccount=1")
- CONFIG_RESOURCE_COUNTERS: enabled
- CONFIG_CGROUP_PERF: enabled
- CONFIG_CFS_BANDWIDTH: enabled
- Storage Drivers:
  - "aufs":
    - CONFIG_AUFS_FS: enabled
    - CONFIG_EXT4_FS_POSIX_ACL: enabled
    - CONFIG_EXT4_FS_SECURITY: enabled
  - "btrfs":
    - CONFIG_BTRFS_FS: enabled
  - "devicemapper":
    - CONFIG_BLK_DEV_DM: enabled
    - CONFIG_DM_THIN_PROVISIONING: enabled
    - CONFIG_EXT4_FS: enabled
    - CONFIG_EXT4_FS_POSIX_ACL: enabled
    - CONFIG_EXT4_FS_SECURITY: enabled
  - "overlay":
    - CONFIG_OVERLAY_FS: missing
    - CONFIG_EXT4_FS_SECURITY: enabled
    - CONFIG_EXT4_FS_POSIX_ACL: enabled
  - "zfs":
    - /dev/zfs: missing
    - zfs command: missing
    - zpool command: missing
```

**Comment by cpuguy83** [comment]
Are you still having an issue here on Docker 1.8 (latest)?
Not sure why name lookup would be failing here.

**Comment by akafred** [comment]
I think we resolved this by using a newer docker-py (1.2.2).
digibib/ls.ext@044803b412e056e1a3cf11d26a3785ecd35e917e

**Comment by thaJeztah** [comment]
Thanks @akafred, I'll close this issue, but feel free to comment if you're still able to reproduce
